### PR TITLE
Don't process blocktxns when we have the block already.

### LINKED ISSUE #9031: Many "peer sent us block transactions for a block we weren't expecting" errors
Currently when a peer received blocks from another peer before a full compact block is received, we get this error when the compact block blocktxns finally arrive. This is because when the block is received MarkBlockAsReceived() is called which removes the mapping of the block being in flight, thereby triggering this error when the blocktxns do arrive, which we requested.

### PR DESCRIPTION
**Pull request by rebroad**
Two checks added. Without this a debug is shown saying the block was "unexpected" but what was actually happening is that another peer is providing the full block before the compact block sequence completes.

Fixes #9031

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -5827,8 +5827,20 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     else if (strCommand == NetMsgType::BLOCKTXN && !fImporting && !fReindex) // Ignore blocks received while importing
     {
         BlockTransactions resp;
+        int nSize = vRecv.size();
         vRecv >> resp;
 
+        BlockMap::iterator mi = mapBlockIndex.find(resp.blockhash);
+        if (mi == mapBlockIndex.end()) {
+            LogPrint("block", "recv blocktxn %s size=%d not recognised peer=%d\n", resp.blockhash.ToString(), nSize, pfrom->id);
+            return true;
```
Comment (TheBlueMatt) [inline src/main.cpp]: This will entirely break downloading of blocks via compact blocks.

Comment (rebroad) [inline src/main.cpp]: How so?

File: src/main.cpp [hunk 2]
```diff
@@ -5827,8 +5827,20 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
     else if (strCommand == NetMsgType::BLOCKTXN && !fImporting && !fReindex) // Ignore blocks received while importing
     {
         BlockTransactions resp;
+        int nSize = vRecv.size();
         vRecv >> resp;
 
+        BlockMap::iterator mi = mapBlockIndex.find(resp.blockhash);
```
Comment (rebroad) [inline src/main.cpp]: Do you mean I need to add a LOCK(cs_main) before this line? I'm still trying to get my head around locks... what is the risk without obtaining a lock first please?

**Comment by TheBlueMatt** [comment]
Oops, misread the line.

**Comment by TheBlueMatt** [comment]
Yea, you need a LOCK(cs_main).

See https://en.m.wikipedia.org/wiki/Lock_(computer_science) for more.

**Comment by rebroad** [comment]
@TheBlueMatt Thanks. Now moved additions to within the LOCK.

**Comment by sdaftuar** [comment]
I don't think there's a situation where this code change would improve the behavior of the node.  It's already the case that if we're not expecting a BLOCKTXN, then we abort processing (in the code just below the code you're adding here).

This change also further complicates the logic, as if somehow we are in the (unlikely) situation where we received a block, pruned it, and then somehow re-requested it via CMPCTBLOCK/GETBLOCKTXN, then this code which returns early if `nTx > 0` without marking the block as received would cause the block download to stall and the peer to be disconnected.

Unless there's some case you have in mind where this patch improves node behavior, I'd say we should leave this alone.

**Pull request closed without merging**